pkg/test/resources/builders/dataplane_builder.go

```diff
@@ -0,0 +1,149 @@
+package builders
+
+import (
+	"context"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/core/resources/store"
+	test_model "github.com/kumahq/kuma/pkg/test/resources/model"
+	"github.com/kumahq/kuma/pkg/util/proto"
+)
+
+var FirstInboundPort = uint32(80)
+var FirstInboundServicePort = uint32(8080)
+var FirstOutboundPort = uint32(10001)
+
+type DataplaneBuilder struct {
+	*core_mesh.DataplaneResource
+}
+
+func Dataplane() *DataplaneBuilder {
+	return &DataplaneBuilder{
+		DataplaneResource: &core_mesh.DataplaneResource{
+			Meta: &test_model.ResourceMeta{
+				Mesh: core_model.DefaultMesh,
+				Name: "dp-1",
+			},
+			Spec: &mesh_proto.Dataplane{
+				Networking: &mesh_proto.Dataplane_Networking{},
+			},
+		},
+	}
+}
+
+func (d *DataplaneBuilder) Build() *core_mesh.DataplaneResource {
+	if err := d.DataplaneResource.Validate(); err != nil {
+		panic(err)
+	}
+	return d.DataplaneResource
+}
+
+func (d *DataplaneBuilder) Create(s store.ResourceStore) error {
+	return s.Create(context.Background(), d.Build(), store.CreateBy(d.Key()))
+}
+
+func (d *DataplaneBuilder) Key() core_model.ResourceKey {
+	return core_model.MetaToResourceKey(d.GetMeta())
+}
+
+func (d *DataplaneBuilder) With(fn func(*core_mesh.DataplaneResource)) *DataplaneBuilder {
+	fn(d.DataplaneResource)
```

Comment:
I could imagine always calling `Validate` here and making sets of changes with `With`.

Reply:
Talked offline. I changed the resource to be hidden in the Builder, so the only way you can get it is by calling `.Build()` which guarantees that it's always valid. Adding validate to `With` seems to be unnecessary, because at any given time a builder can have a resource in an incorrect state. What matters is that you can only receive resource from a builder in the correct state.